src/reader.rs

```diff
@@ -0,0 +1,104 @@
+use std::io::{self, Read};
+use std::cmp::min;
+
+use EntryType;
+use archive::ArchiveInner;
+
+
+pub enum Reader<'a> {
+    Normal(io::Take<&'a ArchiveInner<io::Read + 'a>>),
+    Sparse {
+        data: &'a ArchiveInner<io::Read + 'a>,
+        blocks: Vec<(u64, u64)>,
+        block: usize,
+        position: u64,
+        size: u64,
+    },
+}
+
+impl<'a> Reader<'a> {
+    pub fn new<'x>(typ: EntryType, reader: &'x ArchiveInner<io::Read + 'x>,
+        sparse_chunks: Vec<(u64, u64)>, file_size: u64)
+        -> Reader<'x>
+    {
+        match typ {
+            EntryType::GNUSparse => {
+                Reader::Sparse {
+                    data: reader,
+                    blocks: sparse_chunks,
+                    block: 0,
+                    position: 0,
+                    size: file_size,
+                }
+            }
+            _ => Reader::Normal(reader.take(file_size)),
+        }
+    }
+}
+
+impl<'a> io::Read for Reader<'a> {
+    /// The reader implementation emits sparse file as fully alocated file
+    /// that has absent blocks zeroed
+    ///
+    /// This is non-optimal if you unpack to filesystem but correct and useful
+    /// if you feed the file to some streaming parser or whatever.
+    fn read(&mut self, mut into: &mut [u8]) -> io::Result<usize> {
+        match *self {
+            Reader::Normal(ref mut reader) => reader.read(into),
+            Reader::Sparse {
+                ref mut data, ref blocks, ref mut block,
+                ref mut position, size,
+            } => {
+                let mut bytes_read: usize = 0;
+                while into.len() > bytes_read {
```

Comment:
This seems like it's a pretty complicated loop, perhaps it'd be better expressed as an iterator?

Given an iterator of chunks, it seems like you could `map` that to an iterator of I/O objects like `io::repeat` or the actual underlying archive (alternating).

Perhaps that could simplify some of this?